wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/impl/PersistingProjectRegistry.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.project.server.impl;
+
+import static java.util.Optional.empty;
+import static java.util.stream.Collectors.toSet;
+import static org.eclipse.che.api.fs.server.WsPathUtils.isRoot;
+
+import com.google.common.collect.ImmutableSet;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.model.workspace.config.ProjectConfig;
+import org.eclipse.che.api.fs.server.FsManager;
+import org.eclipse.che.api.fs.server.WsPathUtils;
+import org.eclipse.che.api.project.shared.RegisteredProject;
+import org.eclipse.che.api.project.shared.dto.RegisteredProjectDto;
+import org.eclipse.che.dto.server.DtoFactory;
+
+@Singleton
+public class PersistingProjectRegistry implements ProjectConfigRegistry {
+
+  private final Map<String, RegisteredProject> projects = new ConcurrentHashMap<>();
+
+  private final RegisteredProjectFactory registeredProjectFactory;
+  private final FsManager fsManager;
+  private final FileCache cache;
```

Comment:
I would say it isn't a cache but rather a persistence manager.
Beside this I would add cache (basically a map) not to access the fs for every get operations.

Reply:
@tolusha well, we have this kind of map in default (inmemory, local) registry but an idea is just opposite - to simply reuse a projects volume as a shared for all the machines place (it should be by nature). So it call shared data all the time, not a local map.

So, I'd consider it as a (persisted, not optimized by performance) data cache.

As I said it is experimental and one of the reason is to also learn how it might impact on overall performance. May be it will be OK with this implementation, maybe we can reduce amount of calls, maybe it is not acceptable at all, I do not know yet and would not like to start with complex (data grid like solution) either.

Does it make sense?

Reply:
This makes sense for me, thx.